### PR TITLE
feat(rust, python): preserve base dtype when raising to uint power

### LINKED ISSUE #9051: `pow`: return `Int64` for `Int64` ** `Int64`, and raise if exponent is negative integer?
### Polars version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pypi.org/project/polars/) of Polars.

### Issue description

Currently, `pow` always returns `float` dtype

This came up here: https://github.com/data-apis/dataframe-api/issues/174

The Array API would return `int` for `int` to the power of `int`, and says that `int` to the power of something negative is implementation-dependent

One suggestion could be:
- `int_base_series.pow(int_exponent_series)` returns `int` dtype if `int_exponent_series` is all non-negative
- `int_base_series.pow(int_exponent_series)` raises if `int_exponent_series` contains negative numbers
- `float_base_series.pow(int_exponent_series)` returns `float` dtype
- `int_base_series.pow(float_exponent_series)` returns `float` dtype
- `float_base_series.pow(float_exponent_series)` returns `float` dtype

This is what pyarrow does:
```python

In [18]: pc.power(pa.array([1 ,2]), pa.array([1, 2]))
Out[18]:
<pyarrow.lib.Int64Array object at 0x7f75ea63d000>
[
  1,
  4
]

In [19]: pc.power(pa.array([1 ,2]), pa.array([-1, 2]))
---------------------------------------------------------------------------
ArrowInvalid: integers to negative integer powers are not allowed

In [20]: pc.power(pa.array([1 ,2]), pa.array([-1., 2.]))
Out[20]:
<pyarrow.lib.DoubleArray object at 0x7f75ea6e8460>
[
  1,
  4
]

In [21]: pc.power(pa.array([1. ,2.]), pa.array([-1, 2]))
Out[21]:
<pyarrow.lib.DoubleArray object at 0x7f75ea31f760>
[
  1,
  4
]

In [22]: pc.power(pa.array([1. ,2.]), pa.array([-1., 2.]))
Out[22]:
<pyarrow.lib.DoubleArray object at 0x7f75ea6ebf40>
[
  1,
  4
]
```

### Reproducible example

```python
In [13]: pl.Series([1,2,3]).pow(pl.Series([1,2,3]))
Out[13]:
shape: (3,)
Series: '' [f64]
[
        1.0
        4.0
        27.0
]
```

### Expected behavior

```python
shape: (3,)
Series: '' [i64]
[
        1
        4
        27
]
```
, like `pyarrow`

---

Note: this already was raised [here ](https://github.com/pola-rs/polars/issues/4796), but it looks like there wasn't much discussion, so I'm bringing it up again, hope it's OK

### Installed versions

<details>

```

INSTALLED VERSIONS
------------------
commit           : 37ea63d540fd27274cad6585082c91b1283f963d
python           : 3.10.6.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.10.102.1-microsoft-standard-WSL2
Version          : #1 SMP Wed Mar 2 00:30:59 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_GB.UTF-8
LOCALE           : en_GB.UTF-8

pandas           : 2.0.1
numpy            : 1.24.3
pytz             : 2023.3
dateutil         : 2.8.2
setuptools       : 67.6.1
pip              : 23.0.1
Cython           : None
pytest           : 7.3.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.1.2
IPython          : 8.13.2
pandas_datareader: None
bs4              : 4.12.2
bottleneck       : None
brotli           : None
fastparquet      : None
fsspec           : None
gcsfs            : None
matplotlib       : 3.7.1
numba            : None
numexpr          : None
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : 12.0.0
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : None
snappy           : None
sqlalchemy       : None
tables           : None
tabulate         : None
xarray           : None
xlrd             : None
zstandard        : None
tzdata           : 2023.3
qtpy             : None
pyqt5            : None
None

```

</details>

### PR DESCRIPTION
**Pull request by MarcoGorelli**
closes #9051 

~todo: fixup the return schema for the lazy case~

### CODE REVIEW

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 1]
```diff
@@ -24,15 +24,52 @@ impl Display for PowFunction {
...
+fn pow_on_chunked_arrays<T, F>(
+    base: &ChunkedArray<T>,
+    exponent: &ChunkedArray<F>,
+) -> PolarsResult<Option<Series>>
+where
+    T: PolarsNumericType,
+    F: PolarsNumericType,
+    T::Native: num::pow::Pow<F::Native, Output = T::Native> + ToPrimitive,
+    ChunkedArray<T>: IntoSeries,
+{
```
Comment (MarcoGorelli) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:36]: this is mostly just factored out from the old `pow_on_floats` so it can be shared between the new `pow_on_floats` and the new `pow_on_integral_dtypes`

File: py-polars/tests/unit/test_schema.py [hunk 2]
```diff
@@ -45,16 +45,44 @@ def test_pow_dtype() -> None:
...
     )
-    assert df.collect().dtypes == [pl.UInt32, pl.UInt32, pl.Float64]
+    assert df.collect().dtypes == [
+        pl.UInt32,
+        pl.UInt32,
+        pl.Float64,
+        pl.Int64,
+        pl.UInt64,
+    ]
+    # `scaled_foo2` currently is UInt32 (in the lazy plan) but should be Float64
```
Comment (MarcoGorelli) [inline py-polars/tests/unit/test_schema.py:78]: this issue is unrelated to the current PR:
```python
In [38]: pl.DataFrame({'a': [1, 2, 3]}, schema={'a': pl.UInt32}).lazy().select(pl.col('a')*.5)
Out[38]: <LazyFrame [1 col, {"a": UInt32}] at 0x7F78A110A5C0>

In [39]: pl.DataFrame({'a': [1, 2, 3]}, schema={'a': pl.UInt32}).lazy().select(pl.col('a')*.5).collect()
Out[39]:
shape: (3, 1)
┌─────┐
│ a   │
│ --- │
│ f64 │
╞═════╡
│ 0.5 │
│ 1.0 │
│ 1.5 │
└─────┘
```
Note how the dtype changes after `.collect`

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 3]
```diff
@@ -54,38 +91,79 @@ where
...
+            let base = base.strict_cast(&DataType::Int32)?;
+            let ca = base.i32().unwrap();
+            let exponent = exponent.strict_cast(&DataType::UInt32)?;
+            pow_on_integral_dtypes(ca, exponent.u32().unwrap())
+        },
+        (UInt64, UInt8 | UInt16 | UInt32) => {
+            let ca = base.u64().unwrap();
+            let exponent = exponent.strict_cast(&DataType::UInt32)?;
+            pow_on_integral_dtypes(ca, exponent.u32().unwrap())
+        },
```
Comment (MarcoGorelli) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:152]: right-hand-side - I've put `UInt32`, as 
```rust
use num::pow::Pow;

fn main() {
    let a: i16 = 2;
    let b: u64 = 3;
    println!("a**b: {:?}", Pow::pow(a, b));
}
```
raises with
```
6 |     println!("a**b: {:?}", Pow::pow(a, b));
  |                            --------    ^ the trait `Pow<u64>` is not implemented for `i16`
  |                            |
  |                            required by a bound introduced by this call
```

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 4]
```diff
@@ -54,38 +91,79 @@ where
...
     use DataType::*;
-    match base.dtype() {
-        Float32 => {
+    match (base.dtype(), exponent.dtype()) {
+        (UInt8 | UInt16 | UInt32, UInt8 | UInt16 | UInt32) => {
+            let base = base.strict_cast(&DataType::UInt32)?;
+            let ca = base.u32().unwrap();
+            let exponent = exponent.strict_cast(&DataType::UInt32)?;
+            pow_on_integral_dtypes(ca, exponent.u32().unwrap())
+        },
```
Comment (MarcoGorelli) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:141]: I'm casting these to UInt32 as, trying to pass a ChunkedArray<UInt16> to `pow_on_integral_dtypes` gives
```rust
the trait bound `implementations::SeriesWrap<polars_core::chunked_array::ChunkedArray<UInt16Type>>: polars_core::prelude::SeriesTrait` is not satisfied
the following other types implement trait `polars_core::prelude::SeriesTrait`:
  implementations::SeriesWrap<Logical<DateType, polars_core::datatypes::Int32Type>>
  implementations::SeriesWrap<Logical<DatetimeType, polars_core::datatypes::Int64Type>>
  implementations::SeriesWrap<Logical<DurationType, polars_core::datatypes::Int64Type>>
...
```

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 5]
```diff
@@ -24,15 +24,52 @@ impl Display for PowFunction {
...
+    ChunkedArray<T>: IntoSeries,
+{
+    if (base.len() == 1) && (exponent.len() != 1) {
+        let base = base
+            .get(0)
+            .ok_or_else(|| polars_err!(ComputeError: "base is null"))?;
+
+        Ok(Some(
+            exponent
+                .into_iter()
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:44]: Can we use an `apply` or `apply_on_opt` here. Iterating over chunkedarrays is rather slow.

I understand this might be adapted from code that used the iterator, but I want to move all code away from that, so let's do that on new code as well.

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 6]
```diff
@@ -24,15 +24,52 @@ impl Display for PowFunction {
...
+                .map(|exp| exp.map(|exp| Pow::pow(base, exp)))
+                .collect_trusted::<ChunkedArray<T>>()
+                .into_series(),
+        ))
+    } else {
+        Ok(Some(
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:50]: Can we `allign` chunks and then use a kernel? For the same reason, iterating over chunkedarrays is slow.

File: crates/polars-plan/src/dsl/function_expr/pow.rs (hunk 4 above)
Comment (ritchie46) [reply crates/polars-plan/src/dsl/function_expr/pow.rs:141]: I don't understand? Isn't that because you would send a series otherwise?

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 7]
```diff
@@ -54,38 +91,79 @@ where
...
+    ChunkedArray<T>: IntoSeries,
+{
+    let dtype = T::get_dtype();
+
+    if exponent.len() == 1 {
+        let Some(exponent_value) = exponent.get(0) else {
+            return Ok(Some(Series::full_null(base.name(), base.len(), &dtype)));
+        };
+        let s = match exponent_value.to_u64().unwrap() {
+            a if a == 1 => base.clone().into_series(),
```
Comment (ritchie46) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:116]: Maybe write this somewhat like:

```rust
match a {
1 => 
1..10 => 
_ => 
}
```

File: crates/polars-plan/src/dsl/function_expr/pow.rs (hunk 4 above)
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/pow.rs:141]: sorry I mean that I get that error if I do `exponent.u16()` instead of `exponent.u32()`

File: crates/polars-plan/src/dsl/function_expr/pow.rs (hunk 6 above)
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/pow.rs:54]: thanks for your review! have given this a go

File: py-polars/tests/unit/test_schema.py (hunk 2 above)
Comment (MarcoGorelli) [reply py-polars/tests/unit/test_schema.py:100]: I was just missing `#[cfg(feature = "dtype-u16")]`, solved now

File: crates/polars-plan/src/dsl/function_expr/pow.rs (hunk 4 above)
Comment (MarcoGorelli) [reply crates/polars-plan/src/dsl/function_expr/pow.rs]: I was just missing `#[cfg(feature = "dtype-u16")]`, solved now

File: crates/polars-plan/src/dsl/function_expr/pow.rs [hunk 8]
```diff
@@ -56,34 +85,103 @@ where
...
-        Float32 => {
+    match (base.dtype(), exponent.dtype()) {
+        #[cfg(feature = "dtype-u8")]
+        (UInt8, UInt8 | UInt16 | UInt32 | UInt64) => {
+            let ca = base.u8().unwrap();
+            let exponent = exponent.strict_cast(&DataType::UInt32)?;
+            pow_to_uint_dtype(ca, exponent.u32().unwrap())
+        },
+        #[cfg(feature = "dtype-i8")]
+        (Int8, UInt8 | UInt16 | UInt32 | UInt64) => {
```
Comment (MarcoGorelli) [inline crates/polars-plan/src/dsl/function_expr/pow.rs:139]: hi @ritchie46 - I think you'd mentioned there is a more performant way to do this than strict-casting the exponent to u32

It didn't "register" with me though, but I also didn't press for clarification as you were preparing your presentation
Any chance you could point me to the preferred way of doing this please? Sorry to be asking again, and thanks :pray:

**Review by ritchie46 — APPROVED** [review]

**Comment by corneliusroemer** [comment]
@ritchie46 were you aware that this PR disallows raising a uint series to a negative integer which was previously allowed? I feel it should have been more clearly mentioned in the PR description/title/changelog.

See https://github.com/pola-rs/polars/issues/15897

This issue bit me and was not trivial to debug when upgrading as I had some code like this:

```
priors = (
        pl.scan_ndjson(ndjson, infer_schema_length=10000)
        .select(
            [
                pl.col("nearestNodes"),
                pl.col("nearestNodes")
                .list.len()
                .pow(1).inspect()
                .alias("1/nearestNodesListLength"),
            ]
        )
        .inspect()
        .filter(pl.col("nearestNodes").list.len() > 0)
        .explode("nearestNodes")
        .group_by("nearestNodes")
        .sum()
        .sort("1/nearestNodesListLength", descending=True)
        .with_columns(
            [
                (
                    pl.col("1/nearestNodesListLength")
                    / pl.col("1/nearestNodesListLength").sum()
                ).alias("priors"),
            ]
        )
        .with_columns(
            [
                pl.col("priors").log10().alias("log_priors"),
            ]
        )
        .collect()
    )
```

suddenly giving a very unspecific error:

```
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/opt/homebrew/Caskroom/miniforge/base/envs/py11/lib/python3.11/site-packages/polars/dataframe/frame.py", line 7701, in select
    return self.lazy().select(*exprs, **named_exprs).collect(_eager=True)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/homebrew/Caskroom/miniforge/base/envs/py11/lib/python3.11/site-packages/polars/lazyframe/frame.py", line 1708, in collect
    return wrap_df(ldf.collect())
                   ^^^^^^^^^^^^^
polars.exceptions.ComputeError: conversion from `i32` to `u32` failed in column 'literal' for 1 out of 1 values: [-1]
```

The PR title only talks about raising to `UInt` power, when in fact it _also_ disallows raising to `Int` power.

The docs for pow have not been adjusted but should be: https://docs.pola.rs/docs/python/dev/reference/expressions/api/polars.Expr.pow.html
